source/DataNode.cpp

```diff
@@ -271,6 +272,21 @@ bool DataNode::IsBool(const string &token)
 
 
 
+bool DataNode::IsConditionName(const string &token)
+{
+	// Condition names start with an alphabetic character, and can be followed by spaces, colons, all kinds of brackets,
+	// dashes, single-quotes, alphabetic and numeric characters.
+	static auto validConditionRegex = regex("[[:alpha:]]([-:.\\(\\)\\[\\]<>'\\?,/ [:alnum:]])*");
```

Comment:
This might be a bit too strict. From your description, I gather this would be useful for the AST, but I don't see how that would be the case. We already have individual tokens via DataNode; we can just compare those to a predefined set of special values and operators.

If you are worried about someone trying to give `+` a value, or something similar, we should just blacklist those expressions. For internally reserved names, we could even use a double underscore prefix. I don't want to end up chasing edge cases in plugins where they had a question mark, exclamation mark or asterisk in a mission name that now breaks the game.

Reply:
This is not only for the AST; I've learned in the past that proper input-validation requires that one should compare against allowed characters and an allowed patterns, not against a blacklist of characters.
The difficulty of adding input-validation this late to the codebase could indeed be that some plugins will break because they used an expression that would no longer be allowed.
What we could do is to add warnings for now, but still allow every string of characters to be used as a condition. We could even give a pop-up in-game if a not-allowed condition-name is used.

Reply:
> The difficulty of adding input-validation this late to the codebase could indeed be that some plugins will break because they used an expression that would no longer be allowed.

That's only half of the issue. Let's say I have a mission with an exclamation mark in its name. The game doesn't complain about that, as mission names aren't filtered. The player then completes the mission, and a `mission!: done` condition gets written to the save. Still no warnings. But when the save is loaded, that condition would either give a warning, or be ignored, causing the mission to trigger again. Neither of those is intuitive.

Let's say we try to fix this by filtering mission names. The author of the mission gets a warning, so they rename the mission to something else. But there are players out there who completed the old mission; their save files need to be patched over to the new mission name.

If we don't allow loading conditions with the old mission name, that becomes impossible. If it gives a warning, it can be patched, but the patch still needs to operate on those old conditions forever. In vanilla, we have ancient patches for really old save files, since people do come back to the game after years of not playing. Having these condition patches would block us from actually filtering conditions for quite a few years, if not forever.

Effectively, if we filter conditions this way, we would not only break content, but would also make it impossible to fix.

Of course, we have the option of staying on warnings forever, but that doesn't sound too useful.

Reply:
> That's only half of the issue. Let's say I have a mission with an exclamation mark in its name. The game doesn't complain about that, as mission names aren't filtered.

Agreed, when we validate conditions, then we also need to validate mission names.

> The player then completes the mission, and a `mission!: done` condition gets written to the save. Still no warnings. But when the save is loaded, that condition would either give a warning, or be ignored, causing the mission to trigger again. Neither of those is intuitive.

Agreed, if we do this, then we should filter the mission names using the same rules that apply to conditions.

> Let's say we try to fix this by filtering mission names. The author of the mission gets a warning, so they rename the mission to something else.

Sounds good. The author of the missions should at that point also create a patch to unset the old condition, and set the new condition. (Or comment on the relevant open ticket that the old name should also be supported in the future.)

> But there are players out there who completed the old mission; their save files need to be patched over to the new mission name.

Yes, and for this we need to make sure that there is a lot of time between giving the warnings and starting to perform actual filtering. (Probably measured years.)

> If we don't allow loading conditions with the old mission name, that becomes impossible. If it gives a warning, it can be patched, but the patch still needs to operate on those old conditions forever.

Not forever, a few years should be sufficient.

> In vanilla, we have ancient patches for really old save files, since people do come back to the game after years of not playing. Having these condition patches would block us from actually filtering conditions for quite a few years, if not forever.

Perhaps, but we are now talking about missions and conditions with really uncommon characters in them. How many of such missions do we expect to find?

> Effectively, if we filter conditions this way, we would not only break content, but would also make it impossible to fix.

Not impossible, but we do need to roll-out the filtering really slowly, and start with a very good idea of which names we allow.

> Of course, we have the option of staying on warnings forever, but that doesn't sound too useful.

Like I said; not forever, but I feel we should take a number of steps:

- First we need to agree on the allowed names for missions and conditions. This is going to make a lot people nervous, so it is important that we start with a very good proposal first.
- We create a central tracking ticket, where we explain what is happening, and where people can post comments with names which they feel are okay, but which were rejected. (Allowing us to update the filters to accept things we overlooked.)
- We start with warnings on the command-line for names considered invalid, and we refer to the central tracking ticket in that warning. (Make it easy for people to report names that were rejected but which they feel should be valid.)
- We run with the warnings for over a year, and then we add a second warning in the UI for invalid names (making this visible to all users, beyond developers and power-users).
- We also run in that mode for over a year, and then we should have a clear picture in the central tracking ticket on how many names we are actually rejecting.
- And then we change the filters to actually reject invalid names.

If we use those timescales, then we could still miss plugins that are rarely used, but running for 2 years with those warnings should be sufficient to find rejected names in popular plugins and deal with them.

Changing this PR to draft, since it needs to be changed to command-line warnings only, and we need to describe quite a bit better what we allow and don't allow.
I'll also create this central tracking ticket, when I have a better description of what we accept and don't accept.

Do you agree with this approach?

Reply:
> Not impossible, but we do need to roll-out the filtering really slowly, and start with a very good idea of which names we allow.

Definitely.

> Do you agree with this approach?

Yes, though I'd love to get some input from @warp-core and/or @Amazinite, especially on the timescales.

Reply:
For now let's just only start with a check if conditions, events and missions start with an alphabetic character. This is a quite common convention (also used in C++, and for most person-names in most countries) and I don't need anything more fancy for now.
The bigger discussion (on other validation rules) can then wait for a later moment when we actually make bigger steps on input validation.